cmd/api/api.go

```diff
@@ -187,20 +189,24 @@ func (s *Service) Start() {
 				api.RuntimeFromURLMiddleware(v1BaseURL),
 			},
 			BaseRouter:       baseRouter,
-			ErrorHandlerFunc: api.HumanReadableJsonErrorHandler,
+			ErrorHandlerFunc: api.HumanReadableJsonErrorHandler(*s.logger),
 		})
 	// Manually apply the CORS middleware; we want it to run always.
 	// HandlerWithOptions() above does not apply it to some requests (404 URLs, requests with bad params, etc.).
 	handler = api.CorsMiddleware(handler)
+	// Request context is not cancelled by the server when write timeout is reached. Ensure the context gets canceled.
+	// Ref: https://github.com/golang/go/issues/59602
+	// Metrics middleware should be applied after timeout, since we do not want to cancel the context for metrics.
+	handler = api.ContextTimeoutMiddleware(requestsTimeout)(handler)
 	// Manually apply the metrics middleware; we want it to run always, and at the outermost layer.
 	// HandlerWithOptions() above does not apply it to some requests (404 URLs, requests with bad params, etc.).
 	handler = api.MetricsMiddleware(metrics.NewDefaultRequestMetrics(moduleName), *s.logger)(handler)
 
 	server := &http.Server{
 		Addr:           s.address,
 		Handler:        handler,
-		ReadTimeout:    10 * time.Second,
-		WriteTimeout:   10 * time.Second,
+		ReadTimeout:    requestsTimeout,
+		WriteTimeout:   requestsTimeout,
```

Comment:
ReadTimeout and WriteTimeout are for request I/O and not handling right? these don't have to be the same as the requestsTimeout

Reply:
No, WriteTimeout is not just the IO for writing the response, it includes the server processing.

Reply:
oh, should that be some amount higher than the processing time limit then?

Reply:
Good idea, i think we should; and then by also using the upstream Timeout handler (i mentioned in the other comment) the client will get a 503 in case the processing timeout is reached.

Reply:
Updated.